urlchecker/core/urlproc.py

```diff
@@ -146,13 +149,100 @@ def extract_urls(self):
 
-    def check_urls(self, urls=None, retry_count=1, timeout=5):
+    async def aysnc_url_check(self, url, retry_count, timeout, headers):
+        """
+        Asyncronous check function for one url.
+
+        Args:
+            - urls           (list) : a list of urls to check.
+            - retry_count    (int)  : a number of retries to issue (defaults to 1, no retry).
+            - timeout        (int)  : a timeout in minutes for blocking operations like the connection attempt.
+            - headers        (dict) : headers to use in the request.
+        """
+        # init do retrails and retrails counts
+        do_retry = True
+        rcount = retry_count
+
+        # we will double the time for retry each time
+        retry_seconds = 15
+
+        # With retry, increase timeout by a second
+        pause = timeout
+        saved_responses = []
+        saved_errors = []
+
+        try:
+            while rcount > 0 and do_retry:
+                response = None
+                async with aiohttp.ClientSession(headers=headers) as session:
+                    try:
+                        async with session.get(
+                            url=url,
+                            raise_for_status=False,
+                            timeout=aiohttp.ClientTimeout(pause * 60),
+                        ) as url_response:
+                            response = url_response
+
+                        # if success! Retry is not needed.
+                        if response is not None:
+                            do_retry = False if (response.status == 200) else True
+                            if (response.status == 200) and (rcount == retry_count):
+                                print_success(url)
+                            else:
+                                saved_responses.append(response)
+
+                    except Exception as e:
+                        saved_errors.append(e)
+                        response = None
+
+                # decrement retrials count
+                rcount -= 1
+
+                # If we try again, pause for retry seconds and update retry seconds
+                if rcount > 0 and do_retry:
+                    # keep this only for debugging
+                    await asyncio.sleep(retry_seconds)
+                    retry_seconds = retry_seconds * 2
+                    pause += 1
+
+            if len(saved_errors) > 0 or len(saved_responses) > 0:
+                print_failure(url)
+                # print errors
+                for error_msg in saved_errors:
+                    print("\x1b[33m" + "> " + str(error_msg) + "\x1b[0m")
+
+                # print pevious url checks responses
+                for response_msg in saved_responses[1:]:
+                    check_response_status_code(url, response_msg)
+
+            # When we break from while, we record final response
+            self.record_response(url, response)
+
+        except Exception as e:
```

Comment:
Let's try to not have this extra exception block. We should be aware of issues going on and try to catch them explicitly or otherwise address them.